server/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineResponse.java

```diff
@@ -19,22 +19,94 @@
 
 package org.elasticsearch.action.ingest;
 
+import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentObject;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentParser.Token;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
 public class SimulatePipelineResponse extends ActionResponse implements ToXContentObject {
     private String pipelineId;
     private boolean verbose;
     private List<SimulateDocumentResult> results;
 
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<SimulatePipelineResponse, Void> PARSER =
+        new ConstructingObjectParser<>(
+            "simulate_pipeline_response",
+            true,
+            a -> {
+                List<SimulateDocumentResult> results = (List<SimulateDocumentResult>)a[0];
+                boolean verbose = false;
+                if (results.size() > 0) {
+                    if (results.get(0) instanceof SimulateDocumentVerboseResult) {
+                        verbose = true;
+                    }
+                }
+                return new SimulatePipelineResponse(null, verbose, results);
+            }
+        );
+    static {
+        PARSER.declareObjectArray(
+            constructorArg(),
+            (p, c) -> {
+                ensureExpectedToken(Token.START_OBJECT, p.currentToken(), p::getTokenLocation);
+                boolean isVerbose = false;
+                boolean isFirst = true;
+                SimulateDocumentResult result = null;
+                while (p.nextToken().equals(Token.FIELD_NAME)) {
```

Comment:
I find this loop a bit weird, probably because I am used to a different way of parsing. In a lot of other places we rather do `while ( (token = parser.nextToken()) != END_OBJECT)`. I find that one a bit easier to read as within the while you'd then have the different ifs for the different types of tokens and then you can also reason about naming. This way of parsing also requires us to add the nextToken in the default case, which looks a bit weird too. You would usually skipChildren on a start_array or start_object.

Reply:
Okay. I will try to blend this piece with the other places ;)